### PR TITLE
Remove `friend class` from the library wording

### LINKED ISSUE #5906: [format.arg] remove friend declaration from basic_format_arg::handle
That friend declaration says it's for exposition only, but we don't do the same elsewhere (not even in the adjacent subclauses, e.g. `basic_format_args` needs access to the private data member of `format-arg-store`).

We should remove the friend declaration.

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
Fixes #5906.

### CODE REVIEW

File: source/iostreams.tex [hunk 1]
```diff
@@ -15883,10 +15882,7 @@
 
 \pnum
 \begin{note}
-For purposes of exposition,
-class \tcode{directory_iterator}\iref{fs.class.directory.iterator}
-is shown above as a friend of class \tcode{directory_entry}.
-Friendship allows the \tcode{directory_iterator} implementation to cache
+\tcode{directory_iterator} is allowed to cache
```
Comment (Quuxplusone) [inline source/iostreams.tex:15885]: Are we allowed to say "allowed"?
Should this be turned into a "Recommended practice" note and say something like "Implementations of `directory_iterator` are encouraged to populate attributes of the `directory_entry` object upon construction, without calling `refresh()`"?
OTOH does this note actually mean anything at all? Implementations are _always_ encouraged to omit "unneeded" code; we don't have to tell them that, do we?

File: source/containers.tex [hunk 2]
```diff
@@ -9222,7 +9222,6 @@
 
     // bit reference
     class @\libmember{reference}{vector<bool>}@ {
-      friend class vector;
       constexpr reference() noexcept;
```
Comment (Quuxplusone) [inline source/containers.tex:9225]: This private constructor declaration is probably observable because of something about SFINAE/overload-resolution. But I doubt that its `constexpr` and `noexcept` specifiers are observable.

(Similarly below, the fact that the copy ctor is `=default`ed is unobservable, but perhaps it should be marked `noexcept`. Right now its noexceptness depends on the nothrow-copyability of this class's unseen data members.)

Same comments on `bitset::reference` below.

Comment (jensmaurer) [reply source/containers.tex:9225]: So?  This pull request is not about the private default constructor.

Comment (frederick-vs-ja) [reply source/containers.tex:9225]: > This private constructor declaration is probably observable because of something about SFINAE/overload-resolution. But I doubt that its `constexpr` and `noexcept` specifiers are observable.

Agree. LWG3160 depicted how can its existence be observed.

Currently, LWG4140 is dropping the useless default ctor `bitset<N>::reference`, but still leaves `vector<bool>::reference` untouched. @jwakely should we just extend LWG4140 to modify `vector<bool>::reference` or create a new LWG issue?

File: source/iostreams.tex (hunk 1 above)
Comment (frederick-vs-ja) [reply source/iostreams.tex]: Changed to use "can". (So did #6758.)

File: source/iostreams.tex [hunk 3]
```diff
@@ -16115,10 +16114,7 @@
...
 \pnum
 \begin{note}
-For purposes of exposition,
-class \tcode{directory_iterator}\iref{fs.class.directory.iterator}
-is shown above as a friend of class \tcode{directory_entry}.
-Friendship allows the \tcode{directory_iterator} implementation to cache
+\tcode{directory_iterator} can cache
 already available attribute values
 directly into a \tcode{directory_entry} object
 without the cost of an unneeded call to \tcode{refresh()}.
```
Comment (Quuxplusone) [inline source/iostreams.tex:16120]:
```suggestion
without the cost of a call to \tcode{refresh()}.
```

**Review by Quuxplusone — APPROVED** [review]

File: source/containers.tex (hunk 2 above)
Comment (jwakely) [reply source/containers.tex:9269]: I will add it to the issue, thanks for pointing it out.

File: source/iostreams.tex [hunk 4]
```diff
@@ -16114,7 +16114,7 @@
 
 \pnum
 \begin{note}
-\tcode{directory_iterator} is allowed to cache
+\tcode{directory_iterator} can cache
```
Comment (jensmaurer) [inline source/iostreams.tex:16117]: Please squash this into the earlier commit that is fixed by this change.

Then, force-push.

File: source/iostreams.tex [hunk 5]
```diff
@@ -16117,7 +16117,7 @@
 \tcode{directory_iterator} can cache
 already available attribute values
 directly into a \tcode{directory_entry} object
-without the cost of an unneeded call to \tcode{refresh()}.
+without the cost of a call to \tcode{refresh()}.
```
Comment (jensmaurer) [inline source/iostreams.tex:16120]: This has nothing to do with the topic of this pull request, which is "friend class" removal.

Please move this into a separate pull request and format the change description appropriately.

**Review by jensmaurer — CHANGES_REQUESTED** [review]

File: source/iostreams.tex (hunk 4 above)
Comment (frederick-vs-ja) [reply source/iostreams.tex:16117]: Done.

File: source/iostreams.tex (hunk 3 above)
Comment (frederick-vs-ja) [reply source/iostreams.tex:16120]: Moved the change to #7245.

File: source/iostreams.tex (hunk 5 above)
Comment (frederick-vs-ja) [reply source/iostreams.tex]: Done. Moved the change to #7245.

**Review by jwakely — APPROVED** [review]

**Review by jensmaurer — APPROVED** [review]
